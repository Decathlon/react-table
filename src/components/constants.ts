export const DEFAULT_ROW_HEIGHT = 56;

export const DEFAULT_COLUMN_WIDTH = "auto";

export const DEFAULT_COLSPAN = 1;

export const BORDER_SIZE = 1;

export const MIN_COLUMN_WIDTH = 100;

export const SUBCELL_PADDING = 30;

// must have the same value as max-row-level in style/table/_row.scss
export const MAX_ROW_LEVEL = 2;

export const ROW_SPAN_WIDTH = 110;

export enum MouseClickButtons {
  right = "right",
  left = "left"
}

export const RowHeight = {
  small: 60,
  medium: 80,
  large: 100
};

export const ColumnWidth = {
  small: 90,
  medium: 160,
  large: 230
};
