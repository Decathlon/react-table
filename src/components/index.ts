export { default as ResponsiveContainer } from "./responsive-container";
export * from "./responsive-container";

export { default as Scroller } from "./scroller";
export * from "./scroller";

export { default as Virtualizer } from "./virtualizer";
export * from "./virtualizer";

export * from "./styled-table";
export * from "./table";
export * from "./table-interactions-manager";
export * from "./table-selection";
export * from "./utils";
export * from "./constants";
export * from "./typing";
