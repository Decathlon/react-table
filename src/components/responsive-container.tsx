import * as React from "react";
import classNames from "classnames";

interface IElementSize {
  width: number;
  height: number;
}

export interface IResponsiveContainerOptionalProps {
  className?: string;
}

export interface IResponsiveContainerProps extends IResponsiveContainerOptionalProps {
  children: (size: IElementSize) => JSX.Element;
}

const ResponsiveContainer = ({ className, children }: IResponsiveContainerProps) => {
  const wrapper = React.useRef<HTMLDivElement>(null);
  const [{ height, width }, setSize] = React.useState<IElementSize>({
    height: 0,
    width: 0
  });

  const updateSize = () => {
    if (wrapper && wrapper.current) {
      const { clientWidth, clientHeight } = wrapper.current;
      if (clientWidth !== width || clientHeight !== height) {
        setSize({
          width: clientWidth,
          height: clientHeight
        });
      }
    }
  };

  React.useEffect(() => {
    updateSize();
    window.addEventListener("resize", updateSize);
    return () => {
      window.removeEventListener("resize", updateSize);
    };
  }, [width, height]);

  return (
    <div className={classNames("responsive-container", className)} ref={wrapper}>
      {width && height ? children({ width, height }) : null}
    </div>
  );
};

export default ResponsiveContainer;
