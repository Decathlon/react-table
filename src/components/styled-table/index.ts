export { default as Bubble } from "./bubble";
export * from "./bubble";

export { default as CellWithIcon } from "./cell-with-icon";
export * from "./cell-with-icon";

export { default as HeaderCell } from "./header-cell";
export * from "./header-cell";

export { default as EditableCell } from "./editable-cell";
export * from "./editable-cell";
