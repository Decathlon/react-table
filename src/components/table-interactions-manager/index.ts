export { default as CellDimensionController } from "./cell-dimensions-controller";
export * from "./cell-dimensions-controller";

export { default as ColumnVisibilityController } from "./column-visibility-controller";
export * from "./column-visibility-controller";

export { default as ColumnIdScrollController } from "./column-id-scroll-controller";
export * from "./column-id-scroll-controller";

export { default as FixedColumnController } from "./fixed-column-controller";
export * from "./fixed-column-controller";

export { default as FixedRowController } from "./fixed-row-controller";
export * from "./fixed-row-controller";

export { default as TableInteractionsManager } from "./table-interactions-manager";
export * from "./table-interactions-manager";

export { default as tableManagerReducer } from "./reducers";
export * from "./reducers";
