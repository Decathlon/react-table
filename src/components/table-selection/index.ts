export { default as ContextMenuHandler } from "./context-menu-handler";
export * from "./context-menu-handler";

export { default as SelectionHandler } from "./selection-handler";
export * from "./selection-handler";

export { default as TableSelectionMenu } from "./table-selection-menu";
export * from "./table-selection-menu";
