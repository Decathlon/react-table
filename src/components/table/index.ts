export { default as Table } from "./table";
export * from "./table";

export { default as Row } from "./row";
export * from "./row";

export { default as RowSpan } from "./row-span";
export * from "./row-span";

export { default as Cell } from "./cell";
export * from "./cell";

export { default as ElementaryTable } from "./elementary-table";
export * from "./elementary-table";
