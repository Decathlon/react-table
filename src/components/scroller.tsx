import * as React from "react";
import { isEqual } from "lodash";
import { getScrollbarSize } from "./utils/table";

export const SCROLLBAR_SIZE = getScrollbarSize();

export enum ScrollDirection {
  up = "up",
  down = "down",
  right = "right",
  left = "left",
}

export const VERTICAL_SCROLL_DIRECTIONS = [ScrollDirection.up, ScrollDirection.down];

export const HORIZONTAL_SCROLL_DIRECTIONS = [ScrollDirection.left, ScrollDirection.right];

export enum ScrollOrigin {
  /**  Scrolling is native if it is enabled by the scroll bar */
  native = "native",
  /**  External otherwise */
  external = "external",
}

export interface IOnScroll {
  /**  The scroll direction */
  directions: ScrollDirection[];
  /**  The origin of scrolling (native: scroll bar, external: scrollTo functions) */
  scrollOrigin: ScrollOrigin;
  /**  True if the top of the scroll bar is reached */
  maxTopReached: boolean;
  /**  True if the bottom of the scroll bar is reached */
  maxBottomReached: boolean;
  /**  True if the left of the scroll bar is reached */
  maxLeftReached: boolean;
  /**  True if the right of the scroll bar is reached */
  maxRightReached: boolean;
  /** The vertical scroll value */
  scrollTop: number;
  /** The horizontal scroll value */
  scrollLeft: number;
}

export interface IScrollerProps {
  /** The global content width */
  virtualWidth: number;
  /**  The global content height */
  virtualHeight: number;
  /**  The width of the visible window */
  width: number;
  /**  The height of the visible window */
  height: number;
  /**  The horizontal part width of the scroll bar */
  horizontalPartWidth?: number;
  /** Ignored parts indexes */
  ignoredHorizontalParts?: number[];
  children?: JSX.Element;
  /**  Called when the scroll container has been scrolled */
  onScroll: (scrollValues: IOnScroll) => void;
}

const defaultScroll = { scrollLeft: 0, scrollTop: 0 };

class Scroller extends React.Component<IScrollerProps> {
  private scrollerContainer: React.RefObject<HTMLDivElement> = React.createRef();

  private scrollOrigin: ScrollOrigin = ScrollOrigin.native;

  private scrollTopMax = 0;

  private scrollLeftMax = 0;

  private currentScrollTop = 0;

  private currentScrollLeft = 0;

  private prevIgnoredHorizontalParts?: number[];

  public componentDidMount() {
    this.initializeMaxScrollValues();
  }

  /*
    We update the Scroll component only if his props are modified.
  */
  public shouldComponentUpdate(nextProps: IScrollerProps) {
    return !isEqual(nextProps, this.props);
  }

  public componentDidUpdate(prevProps: IScrollerProps) {
    this.initializeMaxScrollValues();
    const { ignoredHorizontalParts } = this.props;
    if (ignoredHorizontalParts !== prevProps.ignoredHorizontalParts) {
      this.prevIgnoredHorizontalParts = prevProps.ignoredHorizontalParts;
    }
    this.keepHorizontalScrollLeft(prevProps);
  }

  /*
    keep the relative scrollLeft value when the virtualWidth has been changed
  */
  private keepHorizontalScrollLeft = (prevProps: IScrollerProps) => {
    const { virtualWidth, horizontalPartWidth, ignoredHorizontalParts } = this.props;
    if (this.scrollerContainer.current && horizontalPartWidth) {
      const { scrollLeft } = this.scrollerContainer.current;
      if (prevProps.virtualWidth !== virtualWidth && scrollLeft) {
        const diff = prevProps.virtualWidth - virtualWidth;
        const percentageDiff = (100 * diff) / prevProps.virtualWidth;
        const newRelativeDiff = (virtualWidth * percentageDiff) / 100;
        const nbIgnoredHorizontalParts = ignoredHorizontalParts?.length || 0;
        const nbPrevIgnoredHorizontalParts = this.prevIgnoredHorizontalParts?.length || 0;
        const nbRemovedParts = nbPrevIgnoredHorizontalParts - nbIgnoredHorizontalParts;
        // 1 if added or -1 if scrollable parts are removed ;
        const changeDirection = diff >= 0 ? -1 : 1;
        if (nbRemovedParts !== 0) {
          const newLeft = scrollLeft - newRelativeDiff + (changeDirection * newRelativeDiff) / nbRemovedParts;
          this.scrollToLeft(newLeft);
        }
      }
    }
  };

  private initializeMaxScrollValues = () => {
    const { virtualHeight, virtualWidth, width, height } = this.props;
    const hasHorizontalScrollBar = virtualWidth > width;
    const hasVerticalScrollBar = virtualHeight > height;
    const verticalScrollBar = hasVerticalScrollBar ? SCROLLBAR_SIZE : 0;
    const horizontalScrollBar = hasHorizontalScrollBar ? SCROLLBAR_SIZE : 0;
    // We calculate the maximum value of the scroll
    this.scrollTopMax = virtualHeight - height - 5 + horizontalScrollBar;
    // We calculate the maximum value of the scroll left
    this.scrollLeftMax = virtualWidth - width - 5 + verticalScrollBar;
  };

  public getScrollDirections = (): ScrollDirection[] => {
    if (this.scrollerContainer.current) {
      const { scrollTop, scrollLeft } = this.scrollerContainer.current;
      // Return the scroll direction value
      const directions = [];

      if (this.currentScrollTop > scrollTop) {
        directions.push(ScrollDirection.up);
      } else if (this.currentScrollTop < scrollTop) {
        directions.push(ScrollDirection.down);
      }

      if (this.currentScrollLeft > scrollLeft) {
        directions.push(ScrollDirection.left);
      } else if (this.currentScrollLeft < scrollLeft) {
        directions.push(ScrollDirection.right);
      }
      return directions;
    }
    return [];
  };

  public getScrollValues = (): IOnScroll => {
    const currentScrollContainer = this.scrollerContainer.current;
    const { scrollTop, scrollLeft } = currentScrollContainer ? this.scrollerContainer.current || defaultScroll : defaultScroll;
    return {
      directions: this.getScrollDirections(),
      scrollOrigin: this.scrollOrigin,
      maxTopReached: scrollTop <= 0,
      maxBottomReached: scrollTop >= this.scrollTopMax,
      maxLeftReached: scrollLeft <= 0,
      maxRightReached: scrollLeft >= this.scrollLeftMax,
      scrollTop,
      scrollLeft,
    };
  };

  private onContainerScroll = (): void => {
    const { onScroll } = this.props;
    const currentScrollContainer = this.scrollerContainer.current;
    const { scrollTop, scrollLeft } = currentScrollContainer ? this.scrollerContainer.current || defaultScroll : defaultScroll;
    onScroll(this.getScrollValues());
    // Initialize the current scroll's values
    this.currentScrollTop = scrollTop;
    this.currentScrollLeft = scrollLeft;
    // Initialize the scroll's origine to native
    this.scrollOrigin = ScrollOrigin.native;
  };

  public scrollToLeft = (scrollLeft: number) => {
    const { virtualWidth, width } = this.props;
    // We can scroll to left only if the virtualWidth is greater than the width
    if (virtualWidth > width && this.scrollerContainer.current) {
      let newScrollLeft = scrollLeft;
      if (newScrollLeft < 0) {
        newScrollLeft = 0;
      } else if (newScrollLeft > this.scrollLeftMax) {
        newScrollLeft = this.scrollLeftMax;
      }
      if (this.scrollerContainer.current.scrollLeft !== newScrollLeft) {
        this.scrollerContainer.current.scrollLeft = newScrollLeft;
        this.scrollOrigin = ScrollOrigin.external;
        return true;
      }
    }
    return false;
  };

  public scrollToTop = (scrollTop: number) => {
    const { virtualHeight, height } = this.props;
    // We can scroll to top only if the virtualHeight is greater than the height
    if (virtualHeight > height && this.scrollerContainer.current) {
      let newScrollTop = scrollTop;
      if (newScrollTop < 0) {
        newScrollTop = 0;
      } else if (newScrollTop > this.scrollTopMax) {
        newScrollTop = this.scrollTopMax;
      }
      if (this.scrollerContainer.current.scrollTop !== newScrollTop) {
        this.scrollerContainer.current.scrollTop = newScrollTop;
        this.scrollOrigin = ScrollOrigin.external;
        return true;
      }
    }
    return false;
  };

  public render() {
    const { virtualWidth, virtualHeight, width, height, children } = this.props;
    const newChildren = React.Children.map(children, (child) =>
      React.cloneElement(child as React.ReactElement<any>, {
        scrollToLeft: this.scrollToLeft,
        scrollToTop: this.scrollToTop,
      })
    );
    const hasHorizontalScrollBar = virtualWidth > width;
    const hasVerticalScrollBar = virtualHeight > height;
    // We still need a minimum dimension equal to SCROLLBAR_SIZE
    const minHeight = height || SCROLLBAR_SIZE;
    const minWidth = width || SCROLLBAR_SIZE;

    const verticalScrollBar = hasVerticalScrollBar ? SCROLLBAR_SIZE : 0;
    const horizontalScrollBar = hasHorizontalScrollBar ? SCROLLBAR_SIZE : 0;

    return (
      <div
        data-testid="scroller-container"
        className="scroller-container"
        ref={this.scrollerContainer}
        onScroll={this.onContainerScroll}
        style={{
          overflowX: hasHorizontalScrollBar ? "scroll" : "hidden",
          overflowY: hasVerticalScrollBar ? "scroll" : "hidden",
          maxHeight: minHeight,
          maxWidth: minWidth,
        }}
      >
        <div
          className="scroller-content"
          style={{
            minWidth: minWidth - verticalScrollBar,
            minHeight: minHeight - horizontalScrollBar,
          }}
        >
          {newChildren}
        </div>
        <div
          className="scroller-scrollbar"
          style={{
            minWidth: virtualWidth + verticalScrollBar,
            minHeight: virtualHeight + horizontalScrollBar - (hasVerticalScrollBar ? height : 0),
          }}
        />
      </div>
    );
  }
}

export default Scroller;
