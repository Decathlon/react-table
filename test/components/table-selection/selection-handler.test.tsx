/// <reference path="../../typings/tests-entry.d.ts" />
import { createRenderer } from "react-test-renderer/shallow";
import { mount } from "enzyme";

import SelectionHandler from "../../../src/components/table-selection/selection-handler";
import { MouseClickButtons } from "../../../src/components/constants";

describe("SelectionHandler component", () => {
  test("should render children", () => {
    const children = () => <div>Foo</div>;
    const props = {
      children,
    };
    const shallowRenderer = createRenderer();
    shallowRenderer.render(<SelectionHandler {...props} />);
    const rendered = shallowRenderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });

  test("should select cells", () => {
    const source = { rowIndex: 1, cellIndex: 1 };
    const target = { rowIndex: 2, cellIndex: 2 };
    const props = {
      children: jest.fn(),
    };
    const wrapper = mount(<SelectionHandler {...props} />);
    const instance: SelectionHandler = wrapper.instance() as SelectionHandler;
    // @ts-ignore private functions
    instance.onCellMouseDown(source, MouseClickButtons.right);
    expect(instance.state.selectedCells).toEqual({ 1: [1] });
    // @ts-ignore private prop
    expect(instance.startingCell).toEqual(source);
    // @ts-ignore private functions
    instance.onCellMouseEnter(target);
    expect(instance.state.selectedCells).toEqual({ 1: [1, 2], 2: [1, 2] });
  });

  test("should select cells (isDisabledVerticalSelection)", () => {
    const source = { rowIndex: 1, cellIndex: 1 };
    const target = { rowIndex: 2, cellIndex: 2 };
    const props = {
      children: jest.fn(),
      isDisabledVerticalSelection: true,
    };
    const wrapper = mount(<SelectionHandler {...props} />);
    const instance: SelectionHandler = wrapper.instance() as SelectionHandler;
    // @ts-ignore private functions
    instance.onCellMouseDown(source, MouseClickButtons.right);
    expect(instance.state.selectedCells).toEqual({ 1: [1] });
    // @ts-ignore private functions
    instance.onCellMouseEnter(target);
    expect(instance.state.selectedCells).toEqual({ 1: [1, 2] });
  });

  test("should select cells (isDisabledHorizontalSelection)", () => {
    const source = { rowIndex: 1, cellIndex: 1 };
    const target = { rowIndex: 2, cellIndex: 2 };
    const props = {
      children: jest.fn(),
      isDisabledHorizontalSelection: true,
    };
    const wrapper = mount(<SelectionHandler {...props} />);
    const instance: SelectionHandler = wrapper.instance() as SelectionHandler;
    // @ts-ignore private functions
    instance.onCellMouseDown(source, MouseClickButtons.right);
    expect(instance.state.selectedCells).toEqual({ 1: [1] });
    // @ts-ignore private functions
    instance.onCellMouseEnter(target);
    expect(instance.state.selectedCells).toEqual({ 1: [1], 2: [1] });
  });

  test("on cell mouse up", () => {
    const source = { rowIndex: 1, cellIndex: 1 };
    const target = { rowIndex: 2, cellIndex: 2 };
    const props = {
      children: jest.fn(),
      onContextMenu: jest.fn(),
    };
    const wrapper = mount(<SelectionHandler {...props} />);
    const instance: SelectionHandler = wrapper.instance() as SelectionHandler;
    // @ts-ignore private functions
    instance.onCellMouseDown(source, MouseClickButtons.right);
    // @ts-ignore private functions
    instance.onCellMouseEnter(target);
    // @ts-ignore mocked event
    instance.onCellMouseUp();
    // @ts-ignore private prop
    expect(instance.startingCell).toEqual(null);
  });
});
