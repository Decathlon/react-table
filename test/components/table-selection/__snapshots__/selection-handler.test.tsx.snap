// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectionHandler component should render children 1`] = `
<div
  className="selection-handler-container"
>
  <ContextMenuHandler
    selectedCells={{}}
  >
    [Function]
  </ContextMenuHandler>
</div>
`;
